packages/compass-aggregations/src/modules/saved-pipeline.js

```diff
@@ -98,56 +98,60 @@ export const updatePipelineList = () => {
  * @returns {import('redux').AnyAction} The action.
  */
 export const saveCurrentPipeline = () => {
-  return (dispatch, getState) => {
-    const asyncr = require('async');
-    const fs = require('fs');
+  return async (dispatch, getState) => {
+    const { promises: fs } = require('fs');
     const path = require('path');
 
     const state = getState();
 
-    if (state.id === '') {
+    if (getState().id === '') {
       dispatch(createId());
     }
-    const id = getState().id;
 
     const pipeline = state.pipeline.map((stage) => {
       return { ...stage, previewDocuments: [] };
     });
-    track('Aggregation Saved', { id, num_stages: pipeline.length });
-
-    const stateRecord = Object.assign({}
-      , { namespace: state.namespace }
-      , { env: state.env }
-      , { isTimeSeries: state.isTimeSeries }
-      , { isReadonly: state.isReadonly }
-      , { sourceName: state.sourceName }
-      , { pipeline: pipeline }
-      , { name: state.name }
-      , { id: id }
-      , { comments: state.comments }
-      , { sample: state.sample }
-      , { autoPreview: state.autoPreview }
-      , { collation: state.collation }
-      , { collationString: state.collationString }
-    );
 
-    const dirname = getDirectory();
-    asyncr.series([
-      (callback) => {
-        fs.mkdir(dirname, { recursive: true }, () => {
-          callback();
-        });
-      },
-      (callback) => {
-        const fileName = path.join(dirname, `${stateRecord.id}.json`);
-        const options = { encoding: 'utf8', flag: 'w' };
-        fs.writeFile(fileName, JSON.stringify(stateRecord), options, () => {
-          callback(null);
-        });
-      }
-    ], () => {
-      dispatch(updatePipelineList());
+    const {
+      id,
+      name,
+      namespace,
+      comments,
+      sample,
+      autoPreview,
+      collation,
+      collationString,
+      dataService
+    } = getState();
+
+    const stateRecord = {
+      id,
+      name,
+      namespace,
+      comments,
+      sample,
+      autoPreview,
+      collation,
+      collationString,
+      pipeline,
+      host: dataService.dataService.getConnectionString().hosts.join(',')
```

Comment:
This one may not be the most immediately "DE" friendly code, do we want to try/catch this line and make it temporarily optional?

Some more or less related thought for the future:  is `dataService.getConnectionString()` ok in general for connection info or should this rather come from a `currentConnection` state? DataService only knows the part of the connection info that concerns actually connecting (ConnectionOptions) and it can operate without the entire connection info.

Maybe we shouldn't mandate part of connection info to be grabbed from the data service directly, make `getConnectionString` private to Compass or something and have a dedicated ConnectionInfo state that we maintain in the store? Like that we don't have 2 separate source of truth for Connection Infos and we don't have non-serializable stuff like dataService for the state?

Reply:
Added some more optional chaining there so that we don't break if accessing this method is not possible https://github.com/mongodb-js/compass/pull/3186/commits/e5df53a1f18a4c4bd6ef35ed3800db0788f49008